models/ssh_key.go

```diff
@@ -1030,3 +1058,275 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
+	has, err := sess.
+		Where("content = ? AND name = ? AND type = ?", content, name, KeyTypePrincipal).
+		Get(new(PublicKey))
+	if err != nil {
+		return nil, err
+	} else if has {
+		return nil, ErrKeyAlreadyExist{0, "", content}
+	}
+
+	key := &PublicKey{
+		OwnerID:       ownerID,
+		Name:          name,
+		Content:       content,
+		Mode:          AccessModeWrite,
+		Type:          KeyTypePrincipal,
+		LoginSourceID: loginSourceID,
+	}
+	if err = addPrincipalKey(sess, key); err != nil {
+		return nil, fmt.Errorf("addKey: %v", err)
+	}
+
+	return key, sess.Commit()
+}
+
+func addPrincipalKey(e Engine, key *PublicKey) (err error) {
+	// Save SSH key.
+	if _, err = e.Insert(key); err != nil {
+		return err
+	}
+
+	return appendAuthorizedPrincipalsToFile(key)
+}
+
+// appendAuthorizedPrincipalsToFile appends new SSH keys' content to authorized_principals file.
+func appendAuthorizedPrincipalsToFile(keys ...*PublicKey) error {
+	// Don't need to rewrite this file if builtin SSH server is enabled.
+	if setting.SSH.StartBuiltinServer {
+		return nil
+	}
+
+	sshOpLocker.Lock()
+	defer sshOpLocker.Unlock()
+
+	if setting.SSH.RootPath != "" {
+		// First of ensure that the RootPath is present, and if not make it with 0700 permissions
+		// This of course doesn't guarantee that this is the right directory for authorized_keys
+		// but at least if it's supposed to be this directory and it doesn't exist and we're the
+		// right user it will at least be created properly.
+		err := os.MkdirAll(setting.SSH.RootPath, 0700)
+		if err != nil {
+			log.Error("Unable to MkdirAll(%s): %v", setting.SSH.RootPath, err)
+			return err
+		}
+	}
+
+	fPath := filepath.Join(setting.SSH.RootPath, authorizedPrincipalsFile)
+	f, err := os.OpenFile(fPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0600)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Note: chmod is noop on Windows.
+	if !setting.IsWindows {
+		fi, err := f.Stat()
+		if err != nil {
+			return err
+		}
+
+		// .ssh directory should have mode 700, and authorized_keys file should have mode 600.
+		if fi.Mode().Perm() > 0600 {
+			log.Error(authorizedPrincipalsFile+"file has unusual permission flags: %s - setting to -rw-------", fi.Mode().Perm().String())
+			if err = f.Chmod(0600); err != nil {
+				return err
+			}
+		}
+	}
+
+	for _, key := range keys {
+		if _, err = f.WriteString(key.AuthorizedString()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetPrincipalKeyByID returns public key by given ID.
+func GetPrincipalKeyByID(keyID int64) (*PublicKey, error) {
+	return GetPublicKeyByID(keyID)
+}
+
+// CheckPrincipalKeyString checks if the given public key string is recognized by SSH.
+// It returns the actual public key line on success.
+func CheckPrincipalKeyString(content string) (_ string, err error) {
+	if setting.SSH.Disabled {
+		return "", ErrSSHDisabled{}
+	}
+
+	content = strings.TrimRight(content, "\n\r")
+	if strings.ContainsAny(content, "\n\r") {
```

Comment:
Hmm... `\n\r` feels weird - I appreciate that these functions both just look for any characters in any order but it clangs.

Why not just TrimSpace first instead of TrimRight?

```suggestion
	content = strings.TrimSpace(content)
	if strings.ContainsAny(content, "\r\n") {
```

Reply:
This is just copied from the existing functions.
As you can see, the PrincipalKey is almost a copy of PublicKey

Reply:
But trimspace is better